proxylib/memcached/text/parser.go

```diff
@@ -0,0 +1,331 @@
+	tokens := bytes.Fields(data[:linefeed])
+
+	if !reply {
+		meta := meta.MemcacheMeta{
+			Command: string(tokens[0]),
+		}
+		command := tokens[0]
+
+		frameLength := 0
+		noreply := false
+		switch {
+		case p.isCommandRetrieval(command):
+			// get, gets, gat, gats
+			if bytes.HasPrefix(command, []byte("get")) {
+				meta.Keys = tokens[1:]
+			} else if bytes.HasPrefix(command, []byte("gat")) {
+				meta.Keys = tokens[2:]
+			}
+			frameLength = linefeed + 2
+		case p.isCommandStorage(command):
+			// storage commands
+			meta.Keys = tokens[1:2]
+			nBytes, err := strconv.Atoi(string(tokens[4]))
+			if err != nil {
+				log.Error("Failed to parse storage payload length")
+				return proxylib.ERROR, 0
+			}
+			// 2 additional bytes for terminating linefeed
+			frameLength = linefeed + 2 + nBytes + 2
+
+			if command[0] == 'c' { //storage command is "cas"
+				noreply = len(tokens) == CAS_WITH_NOREPLY_FIELDS
+			} else {
+				noreply = len(tokens) == STORAGE_WITH_NOREPLY_FIELDS
+			}
+		case p.isCommandDelete(command):
+			meta.Keys = tokens[1:2]
+			noreply = len(tokens) == DELETE_WITH_NOREPLY_FIELDS
+			frameLength = linefeed + 2
+		case p.isCommandIncrDecr(command):
+			meta.Keys = tokens[1:2]
+			noreply = len(tokens) == INCR_WITH_NOREPLY_FIELDS
+			frameLength = linefeed + 2
+		case bytes.Equal(command, []byte("touch")):
+			meta.Keys = tokens[1:2]
+			noreply = len(tokens) == TOUCH_WITH_NOREPLY_FIELDS
+			frameLength = linefeed + 2
+		case bytes.Equal(command, []byte("slabs")) ||
+			bytes.Equal(command, []byte("lru")) ||
+			bytes.Equal(command, []byte("lru_crawler")) ||
+			bytes.Equal(command, []byte("stats")) ||
+			bytes.Equal(command, []byte("version")) ||
+			bytes.Equal(command, []byte("misbehave")):
+
+			meta.Keys = [][]byte{}
+			frameLength = linefeed + 2
+		case bytes.Equal(command, []byte("flush_all")) ||
+			bytes.Equal(command, []byte("cache_memlimit")):
+			meta.Keys = [][]byte{}
+			noreply = bytes.Equal(tokens[len(tokens)-1], []byte("noreply"))
+			frameLength = linefeed + 2
+		case bytes.Equal(command, []byte("quit")):
+			meta.Keys = [][]byte{}
+			noreply = true
+			frameLength = linefeed + 2
+		case bytes.Equal(command, []byte("watch")):
+			meta.Keys = [][]byte{}
+			frameLength = linefeed + 2
+			p.watching = true
+		default:
+			log.Error("Could not parse text memcache frame")
+			return proxylib.ERROR, 0
+		}
+		logEntry := &cilium.LogEntry_GenericL7{
+			&cilium.L7LogEntry{
+				Proto: "textmemcached",
+				Fields: map[string]string{
+					"command": meta.Command,
+					"keys":    string(bytes.Join(meta.Keys, []byte(", "))),
+				},
+			},
+		}
+
+		p.requestCount++
+		r := &replyIntent{
+			requestID: p.requestCount,
+			command:   command,
+		}
+
+		matches := p.connection.Matches(meta)
+
+		if matches {
+			r.denied = false
+			p.replyQueue = append(p.replyQueue, r)
+			p.connection.Log(cilium.EntryType_Request, logEntry)
+			return proxylib.PASS, frameLength
+		}
+
+		r.denied = true
+		if !noreply {
```

Comment:
would be more readable avoiding the double negative, i.e., maybe refactor to have a variable named `replyExpected`?